LFS-818: Display an estimate of the total number of results in the livetable

IntegralProgrammer opened the pull request:
This PR causes the `PaginationServlet` to check if at least ten times the number of results per LiveTable page match the given query. If the number of matches exceeds this number, the total number of matches is approximated with _"of more than"_, otherwise the exact number of matches is displayed.

For example, suppose that 55 forms match a query. If we are looking at _10_ results per page, the PaginationServlet will check if at least _100_ results can be found. As the amount of found results does _not_ exceed _100_, the _TablePagination_ controls will display `1-10 of 55`, `11-20 of 55`, etc...

Now, suppose that there were 135 matches. The _TablePagination_ controls would display, `1-10 of more than 100`, `11-20 of more than 110`, `21-30 of more than 120`, `31-40 of more than 130`, `41-50 of 135`, `51-60 of 135`, `61-70 of 135`, etc...

To test, build the `LFS-818_test` branch and generate _130_ fake tumor forms with `./generate_fake_tumors.sh`. Test moving between LiveTable pages and applying filters (eg. `Lesion Number > 6500`).

IntegralProgrammer committed:
LFS-818: Display an estimate of the total number of results in the livetable

Check if there are at least 10*page_size more results

marta- reviewed (commented):
`10` appears 3 times in the servlet code. Please define a constant.

The convention that negative means approximation needs to be documented in the Java doc. Alternatively we can make it crystal clear by returning the positive number and adding a boolean field to the json which specifies if the total number is exact or an approximate.

IntegralProgrammer committed:
LFS-818: Display an estimate of the total number of results in the livetable

Replace instances of 10 with a constant, QUERY_SIZE_MULTIPLIER

IntegralProgrammer committed:
LFS-818: Display an estimate of the total number of results in the livetable

Add the isApprox field to the JSON returned from the PaginationServlet and set it true if totalrows is a lower-bound estimate of the total number of query matches. Set it to false if totalrows is an exact value of the number of query matches.

Use the isApprox field as the means to signal that totalrows is an approximate value instead of returning totalrows as a negative value.

IntegralProgrammer commented:
> `10` appears 3 times in the servlet code. Please define a constant.
> 
> The convention that negative means approximation needs to be documented in the Java doc. Alternatively we can make it crystal clear by returning the positive number and adding a boolean field to the json which specifies if the total number is exact or an approximate.

I have replaced the value `10` with the constant `QUERY_SIZE_MULTIPLIER` and have added the field `isApprox` to signal that the `totalrows` value is a lower-bound approximation and not the exact value of the amount of query matches.

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/PaginationServlet.java:
```diff
@@ -484,11 +486,14 @@ private String sanitizeComparator(String comparator)
-        jsonGen.write("totalrows", (limits[3] > limits[1]) ? -1 : (limits[0] + limits[2]));
+        jsonGen.write("totalrows", isApprox
+            ? ((QUERY_SIZE_MULTIPLIER * limits[1]) + limits[0]) : (limits[0] + limits[3]));
+        jsonGen.write("isApprox", isApprox);
```
I suggest an even more descriptive field name:
```suggestion
        jsonGen.write("totalIsApproximate", isApprox);
```

IntegralProgrammer replied:
Implemented by commit `9925415` :heavy_check_mark:

sashaandjic reviewed (approved):
works well

IntegralProgrammer committed:
LFS-818: Display an estimate of the total number of results in the livetable

Rename isApprox to totalIsApproximate
